File: behavioral/C++/ChainOfResponsibility.cpp
#include <iostream>
#include <string>
#include <type_traits>

enum class Planet {
	MERCURY,
	VENUS,
	EARTH,
	MARS,
	JUPITER,
	SATURN,
	URANUS,
	NEPTUNE
};

// Not part of the pattern.
std::string enumToString(Planet planet);

class PlanetHandler
{
protected:
	PlanetHandler* m_Successor = nullptr;

public:
	void setSuccessor(PlanetHandler* successor)
	{
		this->m_Successor = successor;
	}

	virtual void handleRequest(Planet request) = 0;
};

class MercuryHandler : public PlanetHandler
{
public:
	void handleRequest(Planet request) override
	{
		if (request == Planet::MERCURY)
		{
			std::cout << "MercuryHandler handles " << enumToString(request) << std::endl;
			std::cout << "Mercury is hot.\n" << std::endl;
		}
		else
		{
			std::cout << "MercuryHandler doesn't handle " << enumToString(request) << std::endl;
			if (this->m_Successor != nullptr) {
				this->m_Successor->handleRequest(request);
			}
		}
	}
};

class VenusHandler : public PlanetHandler
{
public:
	void handleRequest(Planet request) override
	{
		if (request == Planet::VENUS)
		{
			std::cout << "VenusHandler handles " << enumToString(request) << std::endl;
			std::cout << "Venus is poisonous.\n" << std::endl;
		}
		else
		{
			std::cout << "VenusHandler doesn't handle " << enumToString(request) << std::endl;
			if (this->m_Successor != nullptr) {
				this->m_Successor->handleRequest(request);
			}
		}
	}
};

class EarthHandler : public PlanetHandler
{
public:
	void handleRequest(Planet request) override
	{
		if (request == Planet::EARTH)
		{
			std::cout << "EarthHandler handles " << enumToString(request) << std::endl;
			std::cout << "Earth is comfortable.\n" << std::endl;
		}
		else
		{
			std::cout << "EarthHandler doesn't handle " << enumToString(request) << std::endl;
			if (this->m_Successor != nullptr) {
				this->m_Successor->handleRequest(request);
			}
		}
	}
};

PlanetHandler* setUpChain() {
	PlanetHandler* mercuryHandler = new MercuryHandler();
	PlanetHandler* venusHandler = new VenusHandler();
	PlanetHandler* earthHandler = new EarthHandler();

	mercuryHandler->setSuccessor(venusHandler);
	venusHandler->setSuccessor(earthHandler);

	return mercuryHandler;
}

// Not part of the pattern.
template<typename T>
std::ostream& operator<<(typename std::enable_if<std::is_enum<T>::value, std::ostream>::type& stream, const T& e)
{
	return stream << static_cast<typename std::underlying_type<T>::type>(e);
}

int main()
{
	std::cout << ":: Chain of Responsability ::\n" << std::endl;

	PlanetHandler* chain = setUpChain();
	chain->handleRequest(Planet::VENUS);
	chain->handleRequest(Planet::MERCURY);
	chain->handleRequest(Planet::EARTH);
	chain->handleRequest(Planet::JUPITER);

	system("PAUSE");

	return 0;
}

// Not part of the pattern.
std::string enumToString(Planet planet)
{
	int index = int(planet);

	std::string temp = "";

	switch (index)
	{
	case 0:
		temp = "MERCURY";
		break;
	case 1:
		temp = "VENUS";
		break;
	case 2:
		temp = "EARTH";
		break;
	case 3:
		temp = "MARS";
		break;
	case 4:
		temp = "JUPITER";
		break;
	case 5:
		temp = "SATURN";
		break;
	case 6:
		temp = "URANUS";
		break;
	case 7:
		temp = "NEPTUNE";
		break;
	default:
		break;
	}

	return temp;
}

File: behavioral/javascript/ChainOfResponsibility.js
const Planet = {
  MERCURY: 'MERCURY',
  VENUS: 'VENUS',
  EARTH: 'EARTH',
  MARS: 'MARS',
  JUPITER: 'JUPITER',
  SATURN: 'SATURN',
  URANUS: 'URANUS',
  NEPTUNE: 'NEPTUNE'
};

class PlanetHandler {

  constructor() {
    if (this.constructor === PlanetHandler) {
      throw new TypeError('Cannot construct Abstract instances directly');
    }

    if (this.handleRequest === PlanetHandler.prototype.handleRequest) {
      throw new TypeError('Must override method "handleRequest()"');
    }

    this.successor = null;
  }
  
  setSuccessor(successor) {
    this.successor = successor;
  }

  handleRequest() {
    throw new TypeError('Do not call abstract method "handleRequest()" from child.');
  }

}

class MercuryHandler extends PlanetHandler {

  handleRequest(request) {
    if (request === Planet.MERCURY) {
      console.log("MercuryHandler handles " + request);
      console.log("Mercury is hot.\n");
    } else {
      console.log("MercuryHandler doesn't handle " + request);
      if (this.successor !== null) {
        this.successor.handleRequest(request);
      }
    }
  }

}

class VenusHandler extends PlanetHandler {

  handleRequest(request) {
    if (request === Planet.VENUS) {
      console.log("VenusHandler handles " + request);
      console.log("Venus is poisonous.\n");
    } else {
      console.log("VenusHandler doesn't handle " + request);
      if (this.successor !== null) {
        this.successor.handleRequest(request);
      }
    }
  }

}

class EarthHandler extends PlanetHandler {

  handleRequest(request) {
    if (request === Planet.EARTH) {
      console.log("EarthHandler handles " + request);
      console.log("Earth is comfortable.\n");
    } else {
      console.log("EarthHandler doesn't handle " + request);
      if (this.successor !== null) {
        this.successor.handleRequest(request);
      }
    }
  }

}

function setUpChain() {
  const mercuryHandler = new MercuryHandler();
  const venusHandler = new VenusHandler();
  const earthHandler = new EarthHandler();

  mercuryHandler.setSuccessor(venusHandler);
  venusHandler.setSuccessor(earthHandler);

  return mercuryHandler;
}

console.log(":: Chain of Responsability ::\n");

const chain = setUpChain();
chain.handleRequest(Planet.VENUS);
chain.handleRequest(Planet.MERCURY);
chain.handleRequest(Planet.EARTH);
chain.handleRequest(Planet.JUPITER);